Added code review recipes for Terraform (#380)
microsoft/code-with-engineering-playbook: code-reviews/recipes/Terraform.md

* [ ] Is `try` function used only with simple attribute references and type conversion functions?, as overuse of `try` function to suppress errors will lead to a configuration that is hard to understand and maintain
* [ ] Are the explicit type conversion functions used to normalize types returned only in module outputs?, as the explicit type conversions are rarely necessary in Terraform because it will convert types automatically where required
* [ ] Is `Sensitive` property on schema set to `true` for the fields that contains sensitive information?. This will prevent the field's values from showing up in CLI output

> * [ ] Each defined variables should explicitly have `type information`. E.g. a list(string) or string

Reviewer: ```suggestion
* [ ] Each defined variables should explicitly have type information. E.g. a `list(string)` or `string`
```

Author: Great point.. Change accepted

* [ ] Each defined variables should explicitly have `description` about the purpose of variable and usage

> * [ ] Don’t provide a `default value` for a variable the value of which must be supplied by a user.

Reviewer: ```suggestion
* [ ] Don’t provide a `default` value for a variable the value of which must be supplied by a user.
```

Author: Great point.. Change accepted

> * [ ] Don’t provide a `default` value for a variable the value of which must be supplied by a user.

Reviewer 2: nit: simplify?

 Don’t provide a `default` value for a variable which must be supplied by a user.

Reviewer 2: Also, can we make sure we are consistent using an ending '.' at each list item? Existing list does not have a dot at the end, your items are inconsistent in use.

Author: feedback accepted

Author: and fixed

* [ ] Try avoiding nesting sub configuration within resources. Create a sepearate resource section for resources even though they can be decalred as sub-element of a resource. For example, declaring subnets within virtual network vis-a-vis declaring subnets as a separate resources compared to virtual network on Azure

> * [ ] Never `hard code` any value in configuration. Declare them from locals section if needed multiple times as static value and are internal to the configuration.

Reviewer: ```suggestion
* [ ] Never hard-code any value in configuration. Declare them from locals section if needed multiple times as static value and are internal to the configuration.
```

Author: Great point.. Change accepted

> * [ ] The `names of the resources` created on Azure should not be hard-coded or static. These names should be dynamic and user provided using variables. This is helpful especially in unit testing when multiple tests are running in parallel trying to create resources on Azure but need different names ( few names in azure needs to be named uniquely e.g. storage accounts).

Reviewer: ```suggestion
* [ ] The `name`s of the resources created on Azure should not be hard-coded or static. These names should be dynamic and user-provided using `variable` blocks. This is helpful especially in unit testing when multiple tests are running in parallel trying to create resources on Azure but need different names (few names in Azure need to be named uniquely; e.g. storage accounts).
```

> * [ ] It is a good practice to `output` the `ID of resources` created on Azure from configuration. This is especially helpful when adding dynamic blocks for sub-elements/child elements to the parent resource.

Reviewer: ```suggestion
* [ ] It is a good practice to `output` the ID of resources created on Azure from configuration. This is especially helpful when adding dynamic blocks for sub-elements/child elements to the parent resource.
```

Author: Great point.. Change accepted

* [ ] Use `required_providers` block for establishing the dependency for providers along with pre-determined version.

> * [ ] Use `terraform block` to declare the provider dependency with exact version and also the terraform CLI version needed for the configuration

Reviewer: ```suggestion
* [ ] Use `terraform` block to declare the provider dependency with exact version and also the terraform CLI version needed for the configuration
```

Author: Great point.. Change accepted

> * [ ] Validate the variables values supplied based on usage and type of variable. The validation can be done to variables by adding `validation block`.

Reviewer: ```suggestion
* [ ] Validate the variables values supplied based on usage and type of variable. The validation can be done to variables by adding `validation` block.
```

> * [ ] Validate the variables values supplied based on usage and type of variable. The validation can be done to variables by adding `validation` block.

Reviewer 2: variable's values?

Reviewer 3: feedback accepted

Author: feedback accepted

Author: and fixed
